jetty-client/src/main/java/org/eclipse/jetty/client/util/SPNEGOAuthentication.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    public void setUserKeyTabPath(Path userKeyTabPath)
+    {
+        this.userKeyTabPath = userKeyTabPath;
+    }
+
+    public String getServiceName()
+    {
+        return serviceName;
+    }
+
+    public void setServiceName(String serviceName)
+    {
+        this.serviceName = serviceName;
+    }
+
+    public boolean isUseTicketCache()
+    {
+        return useTicketCache;
+    }
+
+    public void setUseTicketCache(boolean useTicketCache)
+    {
+        this.useTicketCache = useTicketCache;
+    }
+
+    public Path getTicketCachePath()
+    {
+        return ticketCachePath;
+    }
+
+    public void setTicketCachePath(Path ticketCachePath)
+    {
+        this.ticketCachePath = ticketCachePath;
+    }
+
+    public boolean isRenewTGT()
+    {
+        return renewTGT;
+    }
+
+    public void setRenewTGT(boolean renewTGT)
+    {
+        this.renewTGT = renewTGT;
+    }
+
+    @Override
+    public Result authenticate(Request request, ContentResponse response, HeaderInfo headerInfo, Attributes context)
+    {
+        SPNEGOContext spnegoContext = (SPNEGOContext)context.getAttribute(SPNEGOContext.ATTRIBUTE);
+        if (LOG.isDebugEnabled())
+            LOG.debug("Authenticate with context {}", spnegoContext);
+        if (spnegoContext == null)
+        {
+            spnegoContext = login();
+            context.setAttribute(SPNEGOContext.ATTRIBUTE, spnegoContext);
+        }
+
+        String b64Input = headerInfo.getBase64();
+        byte[] input = b64Input == null ? new byte[0] : Base64.getDecoder().decode(b64Input);
+        byte[] output = Subject.doAs(spnegoContext.subject, initGSSContext(spnegoContext, request.getHost(), input));
+        String b64Output = output == null ? null : new String(Base64.getEncoder().encode(output));
+
+        // The result cannot be used for subsequent requests,
+        // so it always has a null URI to avoid being cached.
+        return new SPNEGOResult(null, b64Output);
+    }
+
+    private SPNEGOContext login()
+    {
+        try
+        {
+            // First login via JAAS using the Kerberos AS_REQ call, with a client user.
+            // This will populate the Subject with the client user principal and the TGT.
+            String user = getUserName();
+            if (LOG.isDebugEnabled())
+                LOG.debug("Logging in user {}", user);
+            CallbackHandler callbackHandler = new PasswordCallbackHandler();
+            LoginContext loginContext = new LoginContext("", null, callbackHandler, new SPNEGOConfiguration());
+            loginContext.login();
+            Subject subject = loginContext.getSubject();
+
+            SPNEGOContext spnegoContext = new SPNEGOContext();
+            spnegoContext.subject = subject;
+            if (LOG.isDebugEnabled())
+                LOG.debug("Initialized {}", spnegoContext);
+            return spnegoContext;
+        }
+        catch (LoginException x)
+        {
+            throw new RuntimeException(x);
```

Comment:
Is this logged somewhere? by us? by the users application?

Reply:
This is during the request processing. The exception is reported back to the user as the request is failed.